Basic Shopify Gateway

DanielWright opened the pull request:
This PR takes over for the weirdly bugged #1. It proposes to partially implement Solidus and ActiveMerchant gateways to Shopify. The partial implementation will be sufficient to refund and void payments corresponding to transactions from retail point-of-sale devices.

thisiscab committed:
Add the payment gateway logic and related specs

We add the pos_order_id field to an order, since we can't search for the
order_number with the Spotify API we need to be able to find it by the
ID. Also, I didn't want to use the number field of an order since I feel
like the pos_order_number should belong there.

In this commit, we include the related spree_gateway and active_merchant
gem in order to allow the new shopify gateway to work. Then there is a
single spec (more to come) that make sure that we can successfully
refund an order that was made in Shopify from the Solidus perspective by
using the proper Gateway.

thisiscab committed:
Update Travis so that it contains the Shopify env variables

thisiscab committed:
Change the SHOPIFY_SHOP_NAME value

thisiscab committed:
Fix rubocop offense

Commit:
Preliminary Cleanup of Gemfile

It's strongly preferable to keep dependencies in the gemspec, rather than the
Gemfile, particularly for base dependencies. This allows RubyGems/Bundler to
resolve the dependency tree. The exception is, for obvious reasons, gems not
hosted on RubyGems.

I also removed a development dependency on 'braintree', on which the test suite
doesn't appear to depend. (Maybe it did at one point?)

Commit:
Eliminates Use of Forked Gems

Using custom forks of gems introduces a lot of maintainability headaches down
the road. The juice isn't worth the squeeze: it's simple enough to include the
AM and Solidus gateway implementations directly within our gem.

One complication: @cabouffard's AM gateway was test-driven with Test::Unit. The
unit and integration tests will need to be ported to RSpec.

Commit:
First Attempt at Porting Shopify AM Gateway Unit Tests

There's plenty of opportunity for improvement here. I'm not sure how much I
will contribute, though, as I'm not sure this is a unit test. To be discussed.

thisiscab committed:
WIP - Add unit tests to the shopify refunder

thisiscab committed:
Finish up the tests for the shopify refunder service

thisiscab committed:
Add specs for the shopify voider service

thisiscab committed:
Remove pry from test environment

thisiscab committed:
Namespace the service classes

thisiscab committed:
Dependency inject the required classes

This makes it easier to override if need be and it facilitates testing
those services.

thisiscab committed:
Fix typo

thisiscab committed:
Better use of the RSpec matchers

Instead of checking the trueness of the value, I simply use the
be_success matcher.

thisiscab committed:
Allow the ActiveMerchant Gateway to throw it's own response

Instead of having the voider or refunder throwing an
ActiveMerchant::Billing::Response, the actual Gateway will throw that
reponse depending on the response type received by the voider or
refunder.

thisiscab committed:
Add proper namespacing for general errors and properties

thisiscab committed:
Correctly namespace files / method calls

thisiscab committed:
Refactor the specs + make everything green

thisiscab committed:
Add spec for the refund decorator

thisiscab committed:
Use module convention for specs

Commit:
Resolves Two Spec-Order Dependencies

1. Run in isolation, the Spree::Gateway::ShopifyGateway spec will fail because
   the Spree::Gateway class will not have been loaded.
2. Some specs require the ShopifyAPI be correctly configured, but some specs
   reconfigure it with dummy data. An around(:each) block ensures the original
   configuration is restored after every example is run.

Commit:
Extracts Private Policy Class From ShopifyRefunder

One of the responsibilities taken by ShopifyRefunder is to determine whether
the refund can be issued. I've extracted this responsibility into a private
(for now) CanIssueRefundPolicy object, which can subsequently be separately
unit-tested and promoted as a first-class object.

Commit:
Promotes CanIssueRefundPolicy to First-Class Object

Extracting and promoting the policy object allowed for a significant reduction
in the complexity of the ShopifyRefunder, both implementation and specs. The
refunder now has no knowledge of:

- the conditions under which an error should be raised;
- the relationship between the credit-amount and the transaction-amount.

Commit:
Cleans up ShopifyRefunder Spec

- Consolidates `subject` declarations
- Moves MockError definition closer to invokation for clarity
- Other minor readability tweaks

Commit:
Use Keyword Args in ShopifyRefunder Initializer

With so many arguments, it's greatly preferable to avoid argument-order
dependencies by using keyword arguments. More preferable still would be
reducing the complexity of the object...

Commit:
Delinting

- Separates arguments from injected dependencies for clarity
- Consistent use of symbols for doubles
- Decomposes memoized 'options' into separate arguments

Commit:
Improves Expectations on Refund Creation

The example exercising the creation of the refund didn't set any expectations
on the arguments received by the refund interface. Now it does!

Commit:
Improves Spree::Refund Spec

The spec for 'Refund#pos_order_id' has been improved to exercise its knowledge,
rather than merely its interface.

Commit:
Resolves Demeter Violations in RefundDecorator

Instances of Spree::Payment and Spree::Refund no longer reach through their
associations to retrieve knowledge of their retail order IDs.

Commit:
Fixes Pathname for Application Models

Commit:
Big Ol' Namespace Cleanup

Commit:
Fixes Dummy App Creation

Commit:
Adds RSpec and Rubocop Binstubs

Makes it easier to run either task individually.

Commit:
Travis Debugging

DanielWright commented:
Once #12 is merged, I think this will be ready to merge as well.

Commit:
Revert "Travis Debugging"

This reverts commit 805d239092c6c8b5bdbec8a4024a22cf4e1b5e8a.

Commit:
Obsolesce the Voider (#12)

The `Voider` service object has been rendered obsolete by refactoring the ActiveMerchant and Solidus Shopify gateways to invoke the `Refunder` service object. Some other refactoring, delinting, and other cleanup has also been completed.

DanielWright commented:
I believe this is also ready for a final review (and then merge).

braidn commented:
@DanielWright will take a look through it.

DanielWright commented:
You're an officer and a gentleman and a breeder of fine dogs.

braidn commented on app/policies/spree/retail/shopify/can_issue_refund_policy.rb:
```diff
@@ -0,0 +1,44 @@
+    module Shopify
+      class CanIssueRefundPolicy
+        def initialize(transaction:, amount_to_credit:)
+          @amount_to_credit, @transaction = amount_to_credit, transaction
```
Not a super fan of initialization iVar's like this due to it being somewhat tricky when 'quickly scanning'. Also doesn't Rubocop raise a little warning with these as well?

DanielWright replied:
Fair enough, I can dump the multiple-assignment.

braidn commented on app/services/spree/retail/shopify/refunder.rb:
```diff
@@ -0,0 +1,60 @@
+        @order_id = transaction.order_id
+
+        @refunder_interface = refunder_interface || default_refunder_interface
+        @can_issue_refund_policy_klass = can_issue_refund_policy_klass || default_can_issue_refund_policy_klass
```
Does this read any different with the `can_issue` prefix removed? Is it shorter/less repetitive or harder to discern?

braidn commented on lib/active_merchant/billing/gateways/shopify.rb:
```diff
@@ -0,0 +1,78 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class ShopifyGateway < Gateway
+      self.homepage_url = 'https://shopify.ca/'
```
.ca versus .com?

DanielWright replied:
Either seems to work..? What's the diff? :wink:

braidn commented on lib/active_merchant/billing/gateways/shopify.rb:
```diff
@@ -0,0 +1,78 @@
+      def refund(money, transaction_id, options = {})
+        transaction = find_transaction(transaction_id, options[:order_id])
+
+        return_response create_refund(money, transaction)
```
Is the term 'return_response' what we are actually returning? Not `billing_response` or `transaction_response`?

DanielWright replied:
Hm. I hadn't noticed the implementation of `return_response` (it's a private method below). I think it can be considerably refactored, but fundamentally it returns an `ActiveMerchant::Billing::Response` object with the status of the request and error-messages (if any).

DanielWright replied:
I agree there's a better name to be used, too. :+1:

braidn commented on lib/spree/retail/engine.rb:
```diff
@@ -0,0 +1,28 @@
+      isolate_namespace Spree
+      engine_name 'solidus_retail'
+
+      # TODO(cab): To remove and instead priorize a configuration initializer
```
Maybe we need a `REFACTOR.md` file?

DanielWright replied:
The Dotenv dependency should be removed sooner than later, IMO. We can live with a direct dependency on `ENV` variables, but a gem dependency is a bridge too far. Let's use [`direnv`](http://direnv.net/) and an `.envrc` instead.

DanielWright replied:
I went ahead and did this in 92b1e6f.

braidn commented on spec/models/spree/gateway/shopify_gateway_spec.rb:
```diff
@@ -0,0 +1,92 @@
+    end
+
+    describe "#payment_profiles_supported?" do
+      it { is_expected.to_not be_payment_profiles_supported }
```
Would this read more 'eloquent' using a string to describe the `it` over the shorthand?

DanielWright replied:
```
it "does not support payment profiles" do
  expect(gateway).to_not be_payment_profiles_supported
end
```

Seems like a wash to me.

braidn commented on spec/models/spree/gateway/shopify_gateway_spec.rb:
```diff
@@ -0,0 +1,92 @@
+        let(:payment) { nil }
+
+        it 'raises an exception' do
+          cause = -> { void! }
```
Is this the same as writing `expect { void! }.to raise_error`?

DanielWright replied:
Yes, exactly the same. I like the `expect(&cause)` syntax a bit better when the content of the lambda is more than five characters, but I'm happy to go with the convention that's been established. Would y'all prefer the inline curly-brace syntax?

braidn replied:
Nah, just trying to get a handle on when/when not to use the lambda syntax. No worries. Thanks.

DanielWright replied:
We should probably try to use one syntax, though, yeah?

braidn replied:
We 'should'. This is specifically why I was looking for your mental model around the lambda use: so I could fall in line.

braidn commented:
CR Cred 👀 ! Can we actively avoid these big PR's in the future?

Commit:
Refactoring, Delinting

Following code-review, various small changes are made here:

- renaming of some object parameters and variables
- clearer assignment syntax

DanielWright commented:
I've addressed da CR in 5dd3f2c! :boom:

Commit:
Removes Dotenv Dependency

There are other, cleaner solutions to assigning configuration ENV variables
than introducing a gem dependency.

DanielWright commented:
Shall we get this merged?

braidn commented:
Yep, yep, pls. @cabouffard  also said he was going to be merging his stuff this morning so it's a good time.
